executor/src/witgen/rows.rs

```diff
@@ -34,6 +34,28 @@ impl<T: FieldElement> CellValue<T> {
             _ => Default::default(),
         }
     }
+
+    /// Returns the new combined range constraint or new value for this cell.
+    ///
+    /// # Panics
+    /// Panics if the update is not an improvement.
+    pub fn update_with(&self, c: &Constraint<T>) -> Self {
+        match (self, c) {
+            (CellValue::Known(_), _) => {
+                // Note that this is a problem even if the value that was set is the same,
+                // because we would return that progress was made when it wasn't.
+                panic!("Value was already set.");
+            }
+            (_, Constraint::Assignment(v)) => CellValue::Known(*v),
+            (CellValue::RangeConstraint(current), Constraint::RangeConstraint(c)) => {
+                let new = c.conjunction(current);
+                assert!(new != *current, "Range constraint was already set");
+                log::trace!("         (the conjunction is {})", new);
+                CellValue::RangeConstraint(new)
+            }
+            (_, Constraint::RangeConstraint(c)) => CellValue::RangeConstraint(c.clone()),
```

Comment:
```suggestion
            (CellValue::Unknown, Constraint::RangeConstraint(c)) => {
                CellValue::RangeConstraint(c.clone())
            }
```
Think it's nice if that's explicit.